src/Interpreters/Context.cpp

```diff
@@ -912,6 +912,23 @@ String Context::getFilesystemCacheUser() const
     return shared->filesystem_cache_user;
 }
 
+DatabaseAndTable Context::getOrCacheStorage(const StorageID & id, std::function<DatabaseAndTable()> f, std::optional<Exception> * exception) const
```

Comment:
We could cache it in `Context::resolveStorageID`, but a new method is fine too

Reply:
I don't think we want to enforce this cache everywhere where we call to `Context::resolveStorageID` - at least having a query context is a condition we need to use this cache.

Reply:
We can check this condition in resolveStorageID